File: advent-of-code-25-cs/advent-of-code-25/Puzzles/Puzzle8.cs
using System.Numerics;

namespace advent_of_code_25.Puzzles;

public class Puzzle8 : IPuzzle
{
    private record JunctionBox(long X, long Y, long Z);
    private record Circuits(List<HashSet<JunctionBox>> Items, (JunctionBox, JunctionBox)? LastPair);
    
    private static List<JunctionBox> GetJunctionBoxes()
    {
        const string path = @"InputFiles/input-8";

        return File.ReadLines(path)
            .Select(line => line.Split(',', StringSplitOptions.RemoveEmptyEntries).Select(int.Parse).ToList())
            .Select(numbers => new JunctionBox(numbers[0], numbers[1], numbers[2]))
            .ToList();
    }

    private static double GetDistance(JunctionBox left, JunctionBox right)
    {
        var (p1, p2, p3) = left;
        var (q1, q2, q3) = right;

        return Math.Sqrt(Math.Pow(p1 - q1, 2) + Math.Pow(p2 - q2, 2) + Math.Pow(p3 - q3, 2));
    }

    private static Circuits ConnectCircuitsUntilSingle(List<JunctionBox> boxes, long limit = 1000)
    {
        var sortedPairs = new PriorityQueue<(JunctionBox, JunctionBox), double>();
        var circuits = boxes.Select(box => new HashSet<JunctionBox> {box}).ToList();
        
        for (var i = 0; i < boxes.Count; i++)
        {
            for (var j = i + 1; j < boxes.Count; j++)
            {
                var left = boxes[i];
                var right = boxes[j];
                
                sortedPairs.Enqueue((left, right), GetDistance(left, right));
            }
        }

        (JunctionBox, JunctionBox)? last = null;
        for (var i = 0; i < limit && sortedPairs.Count > 0 && circuits.Count > 1; i++)
        {
            var (left, right) = sortedPairs.Dequeue();
            last = (left, right);
            var leftCircuitIndex = circuits.FindIndex(circuit => circuit.Contains(left));
            var rightCircuitIndex = circuits.FindIndex(circuit => circuit.Contains(right));

            var leftCircuit = circuits[leftCircuitIndex];
            var rightCircuit = circuits[rightCircuitIndex];

            leftCircuit.UnionWith(rightCircuit);
            if (leftCircuitIndex != rightCircuitIndex)
            {
                circuits.RemoveAt(rightCircuitIndex);
            }
        }

        return new Circuits(circuits, last);
    }

    private static long GetCircuitSizes(List<JunctionBox> boxes, long limit = 1000)
    {
        var circuits = ConnectCircuitsUntilSingle(boxes, limit).Items; 
        
        circuits.Sort((left, right) => right.Count - left.Count);
        return circuits.Take(3).Aggregate(1, (current, next) => current * next.Count);
    }

    private static long GetLastPairDistance(List<JunctionBox> boxes)
    {
        var lastPair = ConnectCircuitsUntilSingle(boxes, long.MaxValue).LastPair;
        return (lastPair?.Item1.X ?? 0) * (lastPair?.Item2.X ?? 0);
    }
    
    public void Solution()
    {
        var boxes = GetJunctionBoxes();
        Console.WriteLine(boxes);
        Console.WriteLine(GetCircuitSizes(boxes));
        Console.WriteLine(GetLastPairDistance(boxes));
    }
}

File: advent-of-code-25-ts/src/puzzles/puzzle8.ts
import { createReadStream } from "fs";
import { createInterface } from "readline";
import getFilePath from "../inputFiles/pathHelper.ts";

type JunctionBox = readonly [number, number, number];
interface Pair {
  readonly left: JunctionBox;
  readonly right: JunctionBox;
  readonly distance: number;
}
interface CircuitResult {
  readonly circuits: Set<JunctionBox>[];
  readonly lastPair?: Pair;
}

export default async function Puzzle8() {
  const getJunctionBoxes = async (): Promise<JunctionBox[]> => {
    const path = getFilePath("input-8");
    const fs = createReadStream(path);

    const rl = createInterface({
      input: fs,
      crlfDelay: Infinity,
    });

    const junctionBoxes: JunctionBox[] = [];
    for await (const line of rl) {
      const [x, y, z] = line.split(",", 3).map((el) => parseInt(el ?? "x"));
      if (isNaN(x) || isNaN(y) || isNaN(z)) {
        continue;
      }

      junctionBoxes.push([x, y, z]);
    }

    return junctionBoxes;
  };

  const getDistance = (left: JunctionBox, right: JunctionBox): number => {
    const [p1, p2, p3] = left;
    const [q1, q2, q3] = right;

    return Math.sqrt(Math.pow(p1 - q1, 2) + Math.pow(p2 - q2, 2) + Math.pow(p3 - q3, 2));
  };

  const connectCircuitsUntilSingle = (boxes: JunctionBox[], limit: number): CircuitResult => {
    const sortedPairs: Pair[] = [];
    const circuits: Set<JunctionBox>[] = boxes.map((box) => new Set([box]));

    for (let i = 0; i < boxes.length; i++) {
      for (let j = i + 1; j < boxes.length; j++) {
        const left = boxes[i];
        const right = boxes[j];

        const newPair: Pair = {
          left,
          right,
          distance: getDistance(left, right),
        };

        sortedPairs.push(newPair);
      }
    }
    sortedPairs.sort((left, right) => left.distance - right.distance);

    let lastPair: Pair | undefined = undefined;
    for (let i = 0; i < limit && circuits.length > 1; i++) {
      lastPair = sortedPairs.shift();
      if (lastPair === undefined) {
        break;
      }
      const { left, right } = lastPair;
      const leftCircuitIndex = circuits.findIndex((circuit) => circuit.has(left));
      const rightCircuitIndex = circuits.findIndex((circuit) => circuit.has(right));

      const leftCircuit = circuits[leftCircuitIndex];
      const rightCircuit = circuits[rightCircuitIndex];

      circuits[leftCircuitIndex] = leftCircuit.union(rightCircuit);

      if (leftCircuitIndex !== rightCircuitIndex) {
        circuits.splice(rightCircuitIndex, 1);
      }
    }

    return {
      circuits,
      lastPair,
    };
  };

  const getCircuitSizes = (boxes: JunctionBox[], limit: number): number => {
    const circuits = connectCircuitsUntilSingle(boxes, limit).circuits;

    circuits.sort((left, right) => right.size - left.size);
    return circuits.splice(0, 3).reduce((current, next) => current * next.size, 1);
  };

  const getLastPairDistance = (boxes: JunctionBox[]): number => {
    const lastPair = connectCircuitsUntilSingle(boxes, Number.MAX_SAFE_INTEGER).lastPair;
    return (lastPair?.left[0] ?? 0) * (lastPair?.right[0] ?? 0);
  };

  const boxes = await getJunctionBoxes();
  console.log(getCircuitSizes(boxes, 1000));
  console.log(getLastPairDistance(boxes));
}
